Add additional attributes for redis.search methods create_index, search

mpozniak95 opened the pull request:
# Description

Add additional attributes for redisearch methods: redis.commands.search, create_index:

For create_index method: https://redis-py.readthedocs.io/en/stable/redismodules.html#redis.commands.search.commands.SearchCommands.create_index
Attributes:
- fields - to show from which fields index was created

For search method: https://redis-py.readthedocs.io/en/stable/redismodules.html#redis.commands.search.commands.SearchCommands.search
Attributes:
- query - shows query string
- total - shows number of documents returned
- xdoc{index} - shows returned documents

## Type of change

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added functional tests for these methods

# Does This PR Require a Core Repo Change?

- [ ] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [ ] Followed the style guidelines of this project
- [ ] Changelogs have been updated
- [ ] Unit tests have been added
- [ ] Documentation has been updated

lzchen commented on instrumentation/opentelemetry-instrumentation-redis/src/opentelemetry/instrumentation/redis/util.py:
```diff
@@ -68,3 +68,25 @@ def _format_command_args(args):
+
+
+def _check_skip(name):
+    skip_list = ["FT.SEARCH", "FT.CREATE", "FT.AGGREGATE"]
```
Why are we not tracing these commands in particular?

mpozniak95 replied:
They are traced but with new methods (_traced_search(), etc.). I added that to avoid having repeated spans (one from _traced_execute_command and one from new method). I am not sure if that is the best way to handle it. Maybe we should just keep two spans for those functions or edit current implementation to add those additional attributes to the span?

lzchen replied:
Where is `_traced_search()` defined?

lzchen replied:
Ahh I see, this is an ignore list because these are the commands you are adding more sophisticated tracing for. 

> edit current implementation to add those additional attributes to the span?

In that case I believe this would be the best way to handle this. Utilize the already existing `_traced_execute_command` and create helper methods to determine the span name as well as what attributes should be populated on the span depending on what command it is.

lzchen commented:
@mpozniak95 

I am not too familiar with the inner workings of Redis. Could you update the description of the PR to be more descriptive of why are we making these changes? Is the plan to add INTERNAL spans or should these calls be treated as DATABASE spans?

mpozniak95 commented:
Idea is to integrate it with: https://github.com/traceloop/openllmetry. Since those three methods are probably the most used Redis methods in LLM applications we would like to have additional attributes in spans of that methods.

lzchen commented:
@mpozniak95 

Please update the description of the pr to reflect what you are trying to do. I am not familiar with the inner workings of https://github.com/traceloop/openllmetry so please summarize exactly what you are trying to do and what new methods you are trying to trace (linking to redis docs would be most helpful as well).

> Is the plan to add INTERNAL spans or should these calls be treated as DATABASE spans?

You are currently also creating `INTERNAL` spans for the current implementation. Is this correct or should we treating the spans as `CLIENT` with type database which follows database  [semantic conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/database/redis.md)

xrmx commented on instrumentation/opentelemetry-instrumentation-redis/src/opentelemetry/instrumentation/redis/util.py:
```diff
@@ -68,3 +68,25 @@ def _format_command_args(args):
+
+
+def _set_span_attribute(span, name, value):
+    if value is not None:
```
`if value:` then?

lzchen replied:
Probably still want to keep value if it is `False`.

mpozniak95 commented:
You are right the spans should be `CLIENT` type, I changed it.

mpozniak95 commented:
I changed the implementation to catch those functions in `_traced_execute_command` and edit their span_name and attributes. Also for now I deleted the `aggregate` from the PR, I will add it in separate PR once we merge this.

I feel that it got a little complicated right now, it was easy to extract the response from redis.commands.search function directly (because it was already parsed by `_parse_response `function in redis package) but `execute_command` returns unparsed response. I added some comments to the code (I hope they make any sense, maybe it will be easier to review)

emdneto commented on instrumentation/opentelemetry-instrumentation-redis/src/opentelemetry/instrumentation/redis/__init__.py:
```diff
@@ -230,6 +238,64 @@ def _traced_execute_pipeline(func, instance, args, kwargs):
+        field_attribute = ""
+        # Schema in format:
+        # [first_field_name, first_field_type, first_field_some_attribute1, first_field_some_attribute2, second_field_name, ...]
+        field_types = ["NUMERIC", "TEXT", "GEO", "TAG", "VECTOR"]
```
Can we define a constant for this? And also use list comprehension here?
Something like this:

```
field_attribute = "".join(
    f"Field(name: {schema[index - 1]}, type: {schema[index]});"
    for index in range(1, len(schema))
    if schema[index] in _FIELD_TYPES
)
```

mpozniak95 replied:
Done

emdneto commented on instrumentation/opentelemetry-instrumentation-redis/src/opentelemetry/instrumentation/redis/__init__.py:
```diff
@@ -230,6 +238,64 @@ def _traced_execute_pipeline(func, instance, args, kwargs):
+        _set_span_attribute(
+            span, "redis.search.total", number_of_returned_documents
+        )
+        if "NOCONTENT" in args:
```
do we also need to early return if not number_of_returned_documents?

mpozniak95 replied:
We can, then we don't need another if in line 285.

xrmx commented on instrumentation/opentelemetry-instrumentation-redis/src/opentelemetry/instrumentation/redis/util.py:
```diff
@@ -68,3 +68,16 @@ def _format_command_args(args):
     return out_str
+
+
+def _set_span_attribute(span, name, value):
```
Maybe call this `_set_span_attribute_if_set` to hint that there is a filter?

mpozniak95 replied:
I changed it to `_set_span_attribute_if_value`

lzchen commented on instrumentation/opentelemetry-instrumentation-redis/src/opentelemetry/instrumentation/redis/__init__.py:
```diff
@@ -192,6 +200,10 @@ def _traced_execute_command(func, instance, args, kwargs):
             if callable(request_hook):
                 request_hook(span, instance, args, kwargs)
             response = func(*args, **kwargs)
+            if span.name == "redis.search":
```
I think these should be within the `span.is_recording()` check.

mpozniak95 replied:
Changed

lzchen reviewed (approved):
LGTM. @mpozniak95  Please add a CHANGELOG entry and we can get this merged!

lzchen commented on instrumentation/opentelemetry-instrumentation-redis/src/opentelemetry/instrumentation/redis/__init__.py:
```diff
@@ -181,17 +190,21 @@ def _instrument(
             if callable(request_hook):
                 request_hook(span, instance, args, kwargs)
             response = func(*args, **kwargs)
+            if span.is_recording():
```
Nit: Can't we include this as part of the above "if"?

mpozniak95 replied:
I think no, because one of the additional attributes is based on the response of search method so we need this to be placed after:
`response = func(*args, **kwargs)`
which as I understand should happen even if span is not recording

mpozniak95 commented:
So, can we merge it :)?
